File: api/sync/src/main/java/com/coditory/sherlock/SherlockMigratorBuilder.java
package com.coditory.sherlock;

import com.coditory.sherlock.SherlockMigrator.MigrationChangeSet;
import com.coditory.sherlock.migrator.ChangeSetMethodExtractor;
import com.coditory.sherlock.migrator.MigrationResult;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.coditory.sherlock.Preconditions.expectNonEmpty;
import static com.coditory.sherlock.Preconditions.expectNonNull;

public final class SherlockMigratorBuilder {
    private static final String DEFAULT_MIGRATOR_LOCK_ID = "migrator";
    private final List<MigrationChangeSet> migrationChangeSets = new ArrayList<>();
    private final Set<String> migrationLockIds = new HashSet<>();
    private String migrationId = DEFAULT_MIGRATOR_LOCK_ID;
    private Sherlock sherlock;

    public SherlockMigratorBuilder(Sherlock sherlock) {
        expectNonNull(sherlock, "sherlock");
        this.sherlock = sherlock;
    }

    @NotNull
    public SherlockMigratorBuilder setMigrationId(String migrationId) {
        expectNonEmpty(migrationId, "migrationId");
        ensureUniqueChangeSetId(migrationId);
        this.migrationId = migrationId;
        return this;
    }

    /**
     * Adds change set to migration process.
     *
     * @param changeSetId unique change set id used. This is used as a lock id in migration
     *                    process.
     * @param changeSet   change set action that should be run if change set was not already applied
     * @return the migrator
     */
    @NotNull
    public SherlockMigratorBuilder addChangeSet(String changeSetId, Runnable changeSet) {
        expectNonEmpty(changeSetId, "changeSetId");
        expectNonNull(changeSet, "changeSetId");
        ensureUniqueChangeSetId(changeSetId);
        DistributedLock changeSetLock = createChangeSetLock(changeSetId);
        MigrationChangeSet migrationChangeSet = new MigrationChangeSet(changeSetId, changeSetLock, changeSet);
        migrationChangeSets.add(migrationChangeSet);
        return this;
    }

    @NotNull
    public SherlockMigratorBuilder addAnnotatedChangeSets(@NotNull Object object) {
        expectNonNull(object, "object containing change sets");
        ChangeSetMethodExtractor.extractChangeSets(object, void.class)
                .forEach(changeSet -> addChangeSet(changeSet.getId(), changeSet::execute));
        return this;
    }

    @NotNull
    public SherlockMigrator build() {
        DistributedLock migrationLock = sherlock.createLock()
                .withLockId(migrationId)
                .withPermanentLockDuration()
                .withStaticUniqueOwnerId()
                .build();
        return new SherlockMigrator(migrationLock, migrationChangeSets);
    }

    @NotNull
    public MigrationResult migrate() {
        return build().migrate();
    }

    private DistributedLock createChangeSetLock(String migrationId) {
        return sherlock.createLock()
                .withLockId(migrationId)
                .withPermanentLockDuration()
                .withStaticUniqueOwnerId()
                .build();
    }

    private void ensureUniqueChangeSetId(String changeSetId) {
        if (migrationLockIds.contains(changeSetId)) {
            throw new IllegalArgumentException(
                    "Expected unique change set ids. Duplicated id: " + changeSetId);
        }
        migrationLockIds.add(changeSetId);
    }
}

File: api/coroutines/src/main/java/com/coditory/sherlock/coroutines/migrator/SherlockMigratorBuilder.kt
package com.coditory.sherlock.coroutines.migrator

import com.coditory.sherlock.Preconditions
import com.coditory.sherlock.Preconditions.expectNonEmpty
import com.coditory.sherlock.SherlockDefaults.DEFAULT_MIGRATOR_LOCK_ID
import com.coditory.sherlock.coroutines.DistributedLock
import com.coditory.sherlock.coroutines.Sherlock
import com.coditory.sherlock.coroutines.migrator.SherlockMigrator.MigrationChangeSet
import com.coditory.sherlock.migrator.ChangeSetMethodExtractor
import com.coditory.sherlock.migrator.MigrationChangeSetMethod
import com.coditory.sherlock.migrator.MigrationResult
import java.lang.reflect.InvocationTargetException
import kotlin.coroutines.intrinsics.suspendCoroutineUninterceptedOrReturn

class SherlockMigratorBuilder internal constructor(private val sherlock: Sherlock) {
    private val migrationChangeSets = mutableListOf<MigrationChangeSet>()
    private val migrationLockIds = mutableSetOf<String>()
    private var migrationId = DEFAULT_MIGRATOR_LOCK_ID

    fun setMigrationId(migrationId: String): SherlockMigratorBuilder {
        expectNonEmpty(migrationId, "migrationId")
        ensureUniqueChangeSetId(migrationId)
        this.migrationId = migrationId
        return this
    }

    fun addChangeSet(
        changeSetId: String,
        changeSet: Runnable,
    ): SherlockMigratorBuilder {
        ensureUniqueChangeSetId(changeSetId)
        val changeSetLock = createChangeSetLock(changeSetId)
        val action = { changeSet.run() }
        val migrationChangeSet = MigrationChangeSet(changeSetId, changeSetLock, action)
        migrationChangeSets.add(migrationChangeSet)
        return this
    }

    fun addChangeSet(
        changeSetId: String,
        changeSet: suspend () -> Unit,
    ): SherlockMigratorBuilder {
        ensureUniqueChangeSetId(changeSetId)
        val changeSetLock = createChangeSetLock(changeSetId)
        val migrationChangeSet = MigrationChangeSet(changeSetId, changeSetLock, changeSet)
        migrationChangeSets.add(migrationChangeSet)
        return this
    }

    fun addAnnotatedChangeSets(annotated: Any): SherlockMigratorBuilder {
        Preconditions.expectNonNull(annotated, "object containing change sets")
        ChangeSetMethodExtractor.extractChangeSetMethods(annotated, Void.TYPE)
            .forEach { changeSet ->
                if (ChangeSetMethodExtractor.isCoroutine(changeSet.method())) {
                    addChangeSet(changeSet.id()) { runSuspendedChangeSetMethod(changeSet) }
                } else {
                    addChangeSet(changeSet.id()) { runNonSuspendedChangeSetMethod(changeSet) }
                }
            }
        return this
    }

    private suspend fun runSuspendedChangeSetMethod(changeSet: MigrationChangeSetMethod<Void>) = suspendCoroutineUninterceptedOrReturn<Void> { cont ->
        try {
            changeSet.method.invoke(changeSet.`object`(), cont)
        } catch (e: InvocationTargetException) {
            throw IllegalStateException(
                "Could not invoke suspended changeset method: " + changeSet.method.name,
                e,
            )
        } catch (e: IllegalAccessException) {
            throw IllegalStateException(
                "Could not invoke suspended changeset method: " + changeSet.method.name,
                e,
            )
        }
    }

    private fun runNonSuspendedChangeSetMethod(changeSet: MigrationChangeSetMethod<Void>) {
        try {
            changeSet.method().invoke(changeSet.`object`())
        } catch (e: InvocationTargetException) {
            throw IllegalStateException("Could not invoke changeset method: " + changeSet.method.name, e)
        } catch (e: IllegalAccessException) {
            throw IllegalStateException("Could not invoke changeset method: " + changeSet.method.name, e)
        }
    }

    fun build(): SherlockMigrator {
        val migrationLock =
            sherlock.createLock()
                .withLockId(migrationId)
                .withPermanentLockDuration()
                .withStaticUniqueOwnerId()
                .build()
        return SherlockMigrator(migrationLock, migrationChangeSets)
    }

    suspend fun migrate(): MigrationResult {
        return build().migrate()
    }

    private fun createChangeSetLock(migrationId: String): DistributedLock {
        return sherlock.createLock()
            .withLockId(migrationId)
            .withPermanentLockDuration()
            .withStaticUniqueOwnerId()
            .build()
    }

    private fun ensureUniqueChangeSetId(changeSetId: String) {
        require(!migrationLockIds.contains(changeSetId)) { "Expected unique change set ids. Duplicated id: $changeSetId" }
        migrationLockIds.add(changeSetId)
    }
}
